### PR TITLE
Integrate authorities

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: MSAL/src/public/MSALPublicClientApplication.h [hunk 1]
```diff
@@ -91,7 +92,7 @@
                         not interested in the specific error pass in nil.
  */
 - (id)initWithClientId:(NSString *)clientId
-             authority:(NSString *)authority
+             authority:(MSALAuthority *)authority
```
Comment (oldalton) [inline MSAL/src/public/MSALPublicClientApplication.h]: Let's not replace NSString authority with MSALAuthority in public initializers. Or let's at least deprecate the previous initializer and add a new one.

File: MSAL/src/public/MSALPublicClientApplication.h [hunk 2]
```diff
@@ -31,6 +31,7 @@
 @class MSALResult;
 @class MSALAccount;
 @class MSALTokenRequest;
+@class MSALAuthority;
 
 @interface MSALPublicClientApplication : NSObject
```
Comment (oldalton) [inline MSAL/src/public/MSALPublicClientApplication.h]: getAccounts API will also need to be updated to be asynchronous for multiple authority aliases, but let's take it separately.

File: MSAL/src/requests/MSALInteractiveRequest.m [hunk 3]
```diff
@@ -161,24 +162,41 @@ - (void)run:(MSALCompletionBlock)completionBlock
...
+    if (_parameters.account)
+    {
+        upn = _parameters.account.username;
+    }
+    else if(_parameters.loginHint)
+    {
+        upn = _parameters.loginHint;
+    }
+    
+    [_parameters.unvalidatedAuthority resolveAndValidate:_parameters.validateAuthority
```
Comment (oldalton) [inline MSAL/src/requests/MSALInteractiveRequest.m]: why we don't do resolving in BaseRequest anymore?

File: MSAL/src/requests/MSALSilentRequest.m [hunk 4]
```diff
@@ -66,19 +66,53 @@ - (id)initWithParameters:(MSALRequestParameters *)parameters
 }
 
 - (void)acquireToken:(MSALCompletionBlock)completionBlock
+{
+    NSString *upn = nil;
```
Comment (oldalton) [inline MSAL/src/requests/MSALSilentRequest.m]: this code looks very duplicate of InteractiveRequest. Can we move it to the base request class?

File: MSAL/src/telemetry/MSALTelemetryAPIEvent.h [hunk 5]
```diff
@@ -23,14 +23,12 @@
...
 #import <Foundation/Foundation.h>
 #import "MSALTelemetryApiId.h"
-#import "MSALAuthority.h"
 #import "MSIDTelemetryAPIEvent.h"
 
 @interface MSALTelemetryAPIEvent : MSIDTelemetryAPIEvent
 
 - (void)setUIBehavior:(MSALUIBehavior)uiBehavior;
 - (void)setUser:(MSALAccount *)user;
-- (void)setAuthorityType:(MSALAuthorityType)authorityType;
```
Comment (oldalton) [inline MSAL/src/telemetry/MSALTelemetryAPIEvent.h]: let's still keep set authority type maybe, but pass authority type as string

File: MSAL/src/instance/MSALAuthorityFactory.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import <Foundation/Foundation.h>
+
+@interface MSALAuthorityFactory : NSObject
+
+- (nullable MSALAuthority *)authorityFromUrl:(nonnull NSURL *)url
```
Comment (unpluggedk) [inline MSAL/src/instance/MSALAuthorityFactory.h]: I don't see a need to have this a instance method.
Why not do a class method on these and that'll save us from creating the instance to make these calls?

**Comment by unpluggedk** [comment]
I am fine with merging with or without the sharedInstance on the factory  class. 
If you decide to merge as is, please open another PR or at least file an issue to do it later.

**Pull request merged**